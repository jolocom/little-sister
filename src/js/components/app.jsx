import React from 'react'
import Reflux from 'reflux'
import Radium from 'radium'
import {bankUri} from 'lib/fixtures'

import {Layout, Content} from 'components/layout'
import {
  Paper,
  AppBar,
  IconButton,
  Snackbar,
  FlatButton,
  Dialog} from 'material-ui'
import Badge from 'material-ui/Badge'
import NavigationMenu from 'material-ui/svg-icons/navigation/menu'

import getMuiTheme from 'material-ui/styles/getMuiTheme'

import JolocomTheme from 'styles/jolocom-theme'

import LeftNav from 'components/left-nav/nav.jsx'
import Profile from 'components/accounts/profile.jsx'
import Tour from 'components/tour.jsx'

import Loading from 'components/common/loading.jsx'

import AccountActions from 'actions/account'
import AccountStore from 'stores/account'
import ConfirmStore from 'stores/confirm'

import PinnedActions from 'actions/pinned'
import ConfirmActions from 'actions/confirm'
import ProfileActions from 'actions/profile'
import ProfileStore from 'stores/profile'

import SnackbarStore from 'stores/snackbar'

// A pathname is considered public if either "/" or if it starts
// with any of the following publicRoutes
const publicRoutes = [
  '/login',
  '/signup',
  '/forgot-password',
  '/change-password'
]

let App = React.createClass({

  mixins: [
    Reflux.connect(AccountStore, 'account'),
    Reflux.connect(ProfileStore, 'profile'),
    Reflux.connect(SnackbarStore, 'snackbar'),
    Reflux.connect(ConfirmStore, 'confirm')
  ],

  propTypes: {
    location: React.PropTypes.object,
    children: React.PropTypes.node,
    route: React.PropTypes.object
  },

  contextTypes: {
    router: React.PropTypes.object.isRequired
  },

  childContextTypes: {
    muiTheme: React.PropTypes.object,
    profile: React.PropTypes.any,
    account: React.PropTypes.object,
    username: React.PropTypes.string,
    searchActive: React.PropTypes.bool,
    location: React.PropTypes.object,
    route: React.PropTypes.object
  },

  getChildContext: function () {
    let {account, profile, searchActive} = this.state
    return {
      muiTheme: this.theme,
      profile,
      account,
      username: account && account.username, // backward compat
      searchActive,
      location: this.props.location,
      route: this.props.route
    }
  },

  getInitialState() {
    return {
      searchActive: false,
      searchQuery: null,
      showIndicator: false
    }
  },

  componentWillMount() {
    this.theme = getMuiTheme(JolocomTheme)
    AccountActions.login()
  },

  componentDidUpdate(prevProps, prevState) {
    let {username} = this.state.account

    if (prevState.account.username === undefined ||
      prevState.account.username !== username) {
      this.checkLogin()
    }
  },

  isPublicRoute(path = this.props.location.pathname) {
    return path === '/' ||
      publicRoutes.some((publicRoute) => path.indexOf(publicRoute) === 0)
  },

  checkLogin() {
    let {username, loggingIn} = this.state.account

    // session is still loading, so return for now
    if (username === undefined && loggingIn) {
      return
    }

    if (!username && !this.isPublicRoute()) {
      this.context.router.push('/login')
    } else if (username && this.isPublicRoute()) {
      this.context.router.push('/graph')
    }

    if (username) {
      ProfileActions.load()
    }
  },

  _handlePinnedTap() {
    PinnedActions.show()
  },

  _handleSearchTap() {
    this.refs.search.show()
    this.setState({searchActive: true})
  },

  _handleSearchChange(query) {
    this.setState({searchQuery: query})
  },

  _handleSearchSubmit() {
    let uri = `${bankUri}/${this.state.searchQuery}#this`
    this.context.router.push(`/graph/${encodeURIComponent(uri)}`)
    this.refs.search.hide()
  },

  _handleSearchHide() {
    this.setState({
      searchActive: false,
      searchQuery: null
    })
  },

  _handleChatTap() {
    this.context.router.push('/conversations')
  },

  showDrawer() {
    this.refs.leftNav.show()
  },

  _handleConfirmClose() {
    ConfirmActions.close()
    this.state.confirm.callback()
  },

  getStyles() {
    let styles = {
      container: {
        width: '100%',
        height: '100%',
        position: 'relative'
      },
      header: {
        zIndex: 5,
        backgroundColor: this.theme.appBar.color
      },
      bar: {
        boxShadow: 'none'
      },
      nav: {
        display: this.state.searchActive ? 'none' : 'block'
      },
      icon: {
        color: this.theme.appBar.textColor
      },
      filters: {
        width: '100%',
        height: '48px'
      },
      menuIcon: {
        marginTop: '-20px',
        position: 'relative',
        top: '-4px',
        cursor: 'pointer'
      },
      hamburgerBadge: {
        top: 10,
        right: 20,
        width: 15,
        height: 15,
        display: 'none'
      }
    }
    return styles
  },

  render() {
    const styles = this.getStyles()

    // @TODO render login screen when logging in, also makes sures child
    // components don't get rendered before any user data is available
    if (this.state.account.loggingIn && !this.isPublicRoute()) {
      return <Loading />
    }

    // Deactivating search until we get it working
    /* <IconButton
       iconClassName="material-icons"
       iconStyle={styles.icon}
       onTouchTap={this._handleSearchTap}>search</IconButton> */
    const nav = (
      <div>
        <IconButton
          iconClassName="material-icons"
          iconStyle={styles.icon}
          onTouchTap={this._handleChatTap}>chat</IconButton>
      </div>
    )

    // Deactivating search until we get it working
    /*
    (
      <GraphSearch
        ref="search"
        onChange={this._handleSearchChange}
        onSubmit={this._handleSearchSubmit}
        onHide={this._handleSearchHide} />
    )
    */
    const search = null

    // Deactivating the filters until we get them working
    // <GraphFilters style={styles.filters} showDefaults />
    const filters = null

    const confirmActions = [
      <FlatButton
        label="Cancel"
        primary
        onTouchTap={this._handleConfirmClose}
      />,
      <FlatButton
        label={this.state.confirm.primaryActionText}
        primary
        onTouchTap={this._handleConfirmClose}
      />
    ]

    return (
      <div style={styles.container}>
        {this.isPublicRoute() ? this.props.children : (
          <Layout>
            <Paper style={styles.header}>
              <AppBar
                title="Graph"
                iconElementRight={nav}
                style={styles.bar}
                iconElementLeft={
                  <Badge
                    badgeContent={''}
                    secondary
                    badgeStyle={styles.hamburgerBadge}>
                    <NavigationMenu
                      onTouchTap={this.showDrawer}
                      style={styles.menuIcon} />
                  </Badge>
                }
                onLeftIconButtonTouchTap={this.showDrawer} />
              {filters}
              {search}
            </Paper>
            <LeftNav ref="leftNav" />
            <Content>
              {React.Children.map(this.props.children, (el) => {
                return React.cloneElement(el, {
                  searchQuery: this.state.searchQuery
                })
              })}
            </Content>
            <Profile />
            <Tour />
          </Layout>
        )}

        <Snackbar
          open={this.state.snackbar.open}
          message={this.state.snackbar.message}
          action={this.state.snackbar.undo && 'undo'}
          onActionTouchTap={this.state.snackbar.undoCallback}
        />

        <Dialog
          actions={confirmActions}
          modal={false}
          open={this.state.confirm.open}
          onRequestClose={this.handleClose}
        >
          {this.state.confirm.message}
        </Dialog>
      </div>

    )
  }

})

export default Radium(App)
