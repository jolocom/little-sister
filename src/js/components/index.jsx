import React from 'react'
import Radium from 'radium'
import {RaisedButton} from 'material-ui'
import Carousel from 're-carousel'
import IndicatorDots from 'components/common/indicator-dots.jsx'

let Index = React.createClass({
  contextTypes: {
    router: React.PropTypes.object,
    username: React.PropTypes.string,
    muiTheme: React.PropTypes.object
  },

  componentWillMount() {
    const {account} = this.context
    if (account && account.webId) {
      this.context.router.push('/graph')
    }
  },

  getStyles() {
    let {muiTheme} = this.context

    let styles = {
      container: {
        width: '100%',
        height: '100%',
        position: 'relative',
        display: 'flex',
        flexDirection: 'column',
        backgroundColor: muiTheme.jolocom.gray4
      },
      logo: {
        fontSize: '18px',
        fontWeight: '400',
        textAlign: 'center',
        marginTop: '24px',
        textTransform: 'uppercase'
      },
      logoImg: {
        width: '32px',
        height: '32px',
        verticalAlign: 'middle'
      },
      intro: {
        flex: 1,
        textAlign: 'center'
      },
      slide: {
        backgroundColor: muiTheme.jolocom.gray4,
        padding: '24px',
        height: '100%',
        display: 'flex',
        flexDirection: 'column'
      },
      content: {
        flex: 1
      },
      img: {
        maxWidth: '100%'
      },
      title: {
        color: muiTheme.jolocom.gray1,
        fontSize: '28px',
        fontWeight: '300'
      },
      logoStartImg: {
        padding: '24px',
        maxWidth: '80%',
        width: '320px',
        margin: 'auto',
        marginTop: '20vh'
      },
      actions: {
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'stretch',
        padding: '24px',
        maxWidth: '80%',
        width: '260px',
        margin: 'auto'
      },
      signup: {
        marginBottom: '12px'
      },
      login: {
      }
    }

    return styles
  },

  render() {
    let styles = this.getStyles()

    return (
      <div style={styles.container}>

        {/** <div style={styles.logo}>
          <img src="/img/logo.png" style={styles.logoImg} /> Jolocom
        </div> **/}

        {/** <Carousel style={styles.intro} indicator={IndicatorDots}>
          <div style={styles.slide}>
            <h3 style={styles.title}>Own your data</h3>
            <div style={styles.content}>
              <img src="/img/slide-data.png" style={styles.img} />
            </div>
          </div>
          <div style={styles.slide}>
            <h3 style={styles.title}>Create your personal digital identity</h3>
            <div style={styles.content}>
              <img src="/img/slide-identity.png" style={styles.img} />
            </div>
          </div>
          <div style={styles.slide}>
            <h3 style={styles.title}>Own your data</h3>
            <div style={styles.content}>
              <img src="/img/slide-share.png" style={styles.img} />
            </div>
          </div>
          <div style={styles.slide}>
            <h3 style={styles.title}>Visualize your data</h3>
            <div style={styles.content}>
              <img src="/img/slide-visualize.png" style={styles.img} />
            </div>
          </div>
        </Carousel> **/}

        <div style={styles.intro}>
          <img
            src="/img/logo_littlesister_start.svg"
            style={styles.logoStartImg} />
        </div>

        <div style={styles.actions}>
          <RaisedButton
            secondary
            label="Sign up"
            style={styles.signup}
            onClick={this._handleSignup}
          />
          <RaisedButton
            label="Log in"
            style={styles.login}
            onClick={this._handleLogin}
          />
        </div>
      </div>
    )
  },

  _handleSignup() {
    this.context.router.push('/signup')
  },

  _handleLogin() {
    this.context.router.push('/login')
  }

})

export default Radium(Index)
