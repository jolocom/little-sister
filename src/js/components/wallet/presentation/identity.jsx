import React from 'react'
import Radium from 'radium'

import { TextField, Divider, List, ListItem, Avatar } from 'material-ui'

import Loading from 'components/common/loading'
import CommunicationCall from 'material-ui/svg-icons/communication/call'
import CommunicationEmail from 'material-ui/svg-icons/communication/email'
import CameraIcon from 'material-ui/svg-icons/image/photo-camera'

import {theme} from 'styles'
import {Content, Block} from '../../structure'
import {
  PlusMenu,
  TabContainer,
  HalfScreenContainer,
  ContactList,
  IdCardsList,
  PassportsList,
  InfoDetails
} from './ui'

const STYLES = {
  listItem: {
    whiteSpace: 'nowrap',
    padding: '0 16px 0 72px'
  },
  inputName: {
    color: theme.palette.textColor,
    fontSize: '1em'
  },
  labelName: {
    color: theme.textStyles.labelInputFields.color
  },
  divider: {
    marginLeft: '16px'
  },
  simpleDialog: {
    contentStyle: {
    },
    actionsContainerStyle: {
      textAlign: 'right'
    }
  },
  avatar: {
    marginTop: '10px'
  },
  container: {
    marginLeft: '10px',
    marginRight: '10px'
  }
}

@Radium
export default class WalletIdentity extends React.Component {
  static propTypes = {
    children: React.PropTypes.node,
    username: React.PropTypes.object.isRequired,
    passports: React.PropTypes.array,
    showUserInfo: React.PropTypes.func.isRequired,
    idCards: React.PropTypes.array,
    isLoaded: React.PropTypes.bool.isRequired,
    isError: React.PropTypes.bool.isRequired,
    webId: React.PropTypes.string.isRequired,
    phones: React.PropTypes.array.isRequired,
    emails: React.PropTypes.array.isRequired,
    goToContactManagement: React.PropTypes.func.isRequired,
    goToPassportManagement: React.PropTypes.func.isRequired,
    goToDrivingLicenceManagement: React.PropTypes.func.isRequired,
    onConfirm: React.PropTypes.func.isRequired,
    setFocusedPin: React.PropTypes.func.isRequired,
    changePinValue: React.PropTypes.func.isRequired,
    requestVerificationCode: React.PropTypes.func.isRequired,
    resendVerificationCode: React.PropTypes.func.isRequired,
    enterVerificationCode: React.PropTypes.func.isRequired,
    onVerify: React.PropTypes.func.isRequired,
    saveToBlockchain: React.PropTypes.func.isRequired
  }

  render() {
    const {
      username,
      passports,
      idCards,
      isLoaded,
      webId,
      showUserInfo,
      phones,
      emails,
      goToContactManagement,
      goToPassportManagement,
      goToDrivingLicenceManagement,
      onConfirm,
      changePinValue,
      setFocusedPin,
      requestVerificationCode,
      resendVerificationCode,
      enterVerificationCode,
      saveToBlockchain
    } = this.props

    if (!isLoaded) {
      return <Loading />
    }

    const avatar = (<Avatar
      icon={<CameraIcon viewBox="-3 -3 30 30" />}
      color={theme.jolocom.gray1}
      backgroundColor={theme.jolocom.gray3}
      style={STYLES.avatar} />)

    return (<TabContainer>
      <HalfScreenContainer>
        <Content style={STYLES.container}>
          <Block>
            <List>
              <ListItem
                key={1}
                disabled
                rightIcon={<InfoDetails
                  showDetails={details => showUserInfo(
                    null,
                    details,
                    'OK',
                    STYLES.simpleDialog
                  )}
                  webId={webId}
                  username={username.value} />
                }
                leftAvatar={avatar}
                style={STYLES.listItem}>
                <TextField
                  fullWidth
                  floatingLabelText="Name"
                  inputStyle={STYLES.inputName}
                  floatingLabelStyle={STYLES.labelName}
                  underlineShow={false}
                  floatingLabelFixed
                  value={username.value} />
              </ListItem>
              <Divider style={STYLES.divider} />
            </List>
          </Block>
          <Block>
            <PlusMenu
              name="Contact"
              choice={[...emails, ...phones].length > 0}
              goToManagement={goToContactManagement} />
          </Block>
          <Block>
            <ContactList
              fields={phones}
              changePinValue={changePinValue}
              onConfirm={onConfirm}
              icon={CommunicationCall}
              setFocusedPin={setFocusedPin}
              requestVerificationCode={requestVerificationCode}
              resendVerificationCode={resendVerificationCode}
              enterVerificationCode={enterVerificationCode}
              labelText="Phone Number"
              attrType="phone" />
            <ContactList
              fields={emails}
              onConfirm={onConfirm}
              requestVerificationCode={requestVerificationCode}
              resendVerificationCode={resendVerificationCode}
              enterVerificationCode={enterVerificationCode}
              icon={CommunicationEmail}
              labelText="Email"
              attrType="email" />
          </Block>
          <Block>
            <PlusMenu
              name="Passport"
              choice={passports.length > 0}
              goToManagement={goToPassportManagement} />
          </Block>
          <Block>
            <PassportsList passports={passports} />
          </Block>
          <Block>
            <PlusMenu
              name="ID Card"
              choice={idCards.length > 0}
              goToManagement={goToPassportManagement} />
          </Block>
          <Block>
            <IdCardsList
              idCards={idCards}
              saveToBlockchain={saveToBlockchain} />
          </Block>
          <Block>
            <PlusMenu
              name="Driving License"
              choice={false}
              goToManagement={goToDrivingLicenceManagement} />
          </Block>
        </Content>
      </HalfScreenContainer>
    </TabContainer>)
  }
}
