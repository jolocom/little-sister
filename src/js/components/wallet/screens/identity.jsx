import React from 'react'
import {connect} from 'redux/utils'
import Presentation from '../presentation/identity'
import WalletError from '../presentation/error'

@connect({
  props: ['wallet'],
  actions: [
    'wallet/identity:getIdentityInformation',
    'wallet/identity:changeSmsCodeValue',
    'wallet/identity:changePinValue',
    'wallet/identity:setFocusedPin',
    'wallet/identity:goToPassportManagement',
    'wallet/identity:goToDrivingLicenceManagement',
    'wallet/identity:goToContactManagement',
    'confirmation-dialog:openConfirmDialog',
    'confirmation-dialog:closeConfirmDialog',
    'simple-dialog:showSimpleDialog',
    'simple-dialog:configSimpleDialog',
    'verification:confirmEmail',
    'verification:confirmPhone',
    'verification:startEmailVerification',
    'verification:startPhoneVerification',
    'wallet/identity:saveToBlockchain'
  ]
})

export default class WalletIdentityScreen extends React.Component {
  static propTypes = {
    children: React.PropTypes.node,
    wallet: React.PropTypes.object,
    goToPassportManagement: React.PropTypes.func.isRequired,
    setFocusedPin: React.PropTypes.func.isRequired,
    goToDrivingLicenceManagement: React.PropTypes.func.isRequired,
    goToContactManagement: React.PropTypes.func.isRequired,
    closeConfirmDialog: React.PropTypes.func.isRequired,
    openConfirmDialog: React.PropTypes.func.isRequired,
    getIdentityInformation: React.PropTypes.func.isRequired,
    configSimpleDialog: React.PropTypes.func.isRequired,
    showSimpleDialog: React.PropTypes.func.isRequired,
    startPhoneVerification: React.PropTypes.func.isRequired,
    startEmailVerification: React.PropTypes.func.isRequired,
    confirmEmail: React.PropTypes.func.isRequired,
    confirmPhone: React.PropTypes.func.isRequired,
    resendVerificationLink: React.PropTypes.func,
    resendVerificationSms: React.PropTypes.func,
    changePinValue: React.PropTypes.func.isRequired,
    changeSmsCodeValue: React.PropTypes.func.isRequired,
    saveToBlockchain: React.PropTypes.func.isRequired
  }

  componentWillMount() {
    this.props.getIdentityInformation()
  }

  render() {
    const {username, contact, webId, passports, idCards, loaded, error
    } = this.props.wallet.identity
    if (error) {
      return (<WalletError
        message="...oops something went wrong! We were not able to load your data." // eslint-disable-line max-len
        buttonLabel="RETRY"
        onClick={() => this.render()} />)
    }
    const {emails, phones} = contact

    return (<Presentation
      username={username}
      emails={emails}
      phones={phones}
      webId={webId}
      passports={passports}
      idCards={idCards}
      isLoaded={loaded}
      isError={error}
      setFocusedPin={this.props.setFocusedPin}
      changePinValue={this.props.changePinValue}
      goToContactManagement={this.props.goToContactManagement}
      goToPassportManagement={this.props.goToPassportManagement}
      goToDrivingLicenceManagement={this.props.goToDrivingLicenceManagement}
      requestIdCardVerification={
        ({message, rightButtonLabel, leftButtonLabel, index}) =>
          this.props.openConfirmDialog(
            message,
            rightButtonLabel,
            () => { this.props.saveToBlockchain(0) },
            leftButtonLabel
          )
        }
      requestVerificationCode={(args, params) => this.showVerificationWindow(args, () => { // eslint-disable-line max-len
        return () => this.showVerificationWindow(params,
          (callbackArgs) => this.requestVerificationCode(callbackArgs))
      })}
      resendVerificationCode={(...args) => this.showVerificationWindow(...args,
        (...params) => this.resendVerificationCode(...params)
      )}
      enterVerificationCode={(...args) => this.showVerificationWindow(...args,
        (...params) => this.enterVerificationCode(...params)
      )}
      onVerify={({title, message, buttonText, style, attrValue}) => {
        this.props.configSimpleDialog(() => {
          this.props.startEmailVerification({email: attrValue})
        }, title, message, buttonText, style)
        this.props.showSimpleDialog()
      }}
      onConfirm={(...args) => { this.onConfirm(...args) }}
      showUserInfo={(...args) => {
        this.props.configSimpleDialog(...args)
        this.props.showSimpleDialog()
      }} />)
  }

  onConfirm(args, params) {
    return this.showVerificationWindow(args, () => {
      return () => this.showVerificationWindow(params,
      (callbackArgs) => this.requestVerificationCode(callbackArgs))
    })
  }

  requestVerificationCode({attrType, attrValue, index}) {
    if (attrType === 'phone') {
      return () => this.props.startPhoneVerification({phone: attrValue, index})
    } else if (attrType === 'email') {
      return () => this.props.startEmailVerification({email: attrValue, index})
    }
  }

  resendVerificationCode({attrType, attrValue, index}) {
    if (attrType === 'phone') {
      return () => this.props.resendVerificationLink({phone: attrValue, index})
    } else if (attrType === 'email') {
      return () => this.props.resendVerificationSms({email: attrValue, index})
    }
  }

  enterVerificationCode({attrType, attrValue = 'attrValue'}) {
    if (attrType === 'phone') {
      return () => this.props.confirmPhone({phone: attrValue})
    } else if (attrType === 'email') {
      return () => this.props.confirmEmail({email: attrValue})
    }
  }

  showVerificationWindow({title, message, attrValue, attrType, index, rightButtonLabel, leftButtonLabel}, callback) { // eslint-disable-line max-len
    return this.props.openConfirmDialog(
      title,
      message,
      rightButtonLabel,
      callback({attrValue, attrType, index}),
      leftButtonLabel
    )
  }
}
