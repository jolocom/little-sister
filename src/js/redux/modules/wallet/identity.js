import Immutable from 'immutable'
import { makeActions } from '../'
import * as router from '../router'
import WebIdAgent from 'lib/agents/webid'

const actions = module.exports = makeActions('wallet/identity', {
  goToContactManagement: {
    expectedParams: [],
    creator: () => {
      return (dispatch) => {
        dispatch(router.pushRoute('/wallet/identity/contact'))
      }
    }
  },
  changeSmsCodeValue: {
    expectedParams: ['value', 'index']
  },
  changePinValue: {
    expectedParams: ['value', 'index']
  },
  setFocusedPin: {
    expectedParams: ['value', 'index']
  },
  goToPassportManagement: {
    expectedParams: [],
    creator: () => {
      return (dispatch) => {
        dispatch(router.pushRoute('/wallet/identity/passport/add'))
      }
    }
  },
  goToDrivingLicenceManagement: {
    expectedParams: [],
    creator: () => {
      return (dispatch) => {
        dispatch(router.pushRoute('/wallet/identity/drivers-licence/add'))
      }
    }
  },
  goToIdentity: {
    expectedParams: [],
    creator: () => {
      return (dispatch) => {
        dispatch(router.pushRoute('/wallet/identity/'))
      }
    }
  },
  getIdentityInformation: {
    expectedParams: [],
    async: true,
    creator: (params) => {
      return (dispatch, getState, {services, backend}) => {
        dispatch(actions.getIdentityInformation.buildAction(params, () => {
          return backend.solid.getUserInformation(new WebIdAgent().getWebId())
        }))
      }
    }
  }
})

const mapBackendToState = ({webId, contact, passports, idCards}) =>
  Immutable.fromJS({
    loaded: true,
    error: false,
    webId,
    username: {value: webId.split('.')[0].split('://')[1]},
    contact: {
      emails: contact.email,
      phones: contact.phone
    },
    passports: passports,
    idCards: idCards
  })
const mapBackendToStateError =
({webId, username, contact, passports, idCards}) =>
  Immutable.fromJS({
    loaded: true,
    error: true,
    webId: {value: ''},
    username: {value: ''},
    contact: {
      emails: [],
      phones: []
    },
    passports: [],
    idCards: []
  })
const initialState = Immutable.fromJS({
  loaded: false,
  error: false,
  webId: '',
  username: {
    verified: false,
    value: ''
  },
  contact: {
    phones: [{
      type: '',
      number: '',
      verified: false,
      smsCode: '',
      pin: '',
      pinFocused: false
    }],
    emails: [{
      type: '',
      address: '',
      pin: '',
      verified: false
    }]
  },
  passports: [
    {
      number: '',
      givenName: '',
      familyName: '',
      birthDate: '',
      gender: '',
      showAddress: '',
      streetAndNumber: '',
      city: '',
      zip: '',
      state: '',
      country: '',
      verified: false
    }
  ]
})

const changeSmsCodeValue = (state, {index, value}) => {
  if (/^[0-9]{0,6}$/.test(value)) {
    return state.mergeIn(['contact', 'phones', index], {
      smsCode: value
    })
  }
  return state
}

const changePinValue = (state, {index, value}) => {
  if (/^[0-9]{0,6}$/.test(value)) {
    return state.mergeIn(['contact', 'phones', index], {
      pin: value
    })
  }
  return state
}

module.exports.default = (state = initialState, action = {}) => {
  switch (action.type) {
    case actions.getIdentityInformation.id_success:
      return mapBackendToState(action.result)

    case actions.getIdentityInformation.id_fail:
      return mapBackendToStateError(state)

    case actions.changeSmsCodeValue.id:
      return changeSmsCodeValue(state, action)

    case actions.changePinValue.id:
      return changePinValue(state, action)

    case actions.setFocusedPin.id:
      return state.mergeIn(['contact', 'phones', action.index], {
        pinFocused: action.value
      })
    default:
      return state
  }
}
