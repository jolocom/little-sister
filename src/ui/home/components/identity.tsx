import React from 'react'
import Icon from 'react-native-vector-icons/MaterialCommunityIcons'
import { TextInput, StyleSheet, TouchableOpacity } from 'react-native'
import { Container, Block } from 'src/ui/structure'
import { JolocomTheme } from 'src/styles/jolocom-theme'
import { QRcodeScanner } from 'src/ui/home/components/qrcodeScanner'

// TODO Type for E, Event is not enough
interface Props {
  scanning: boolean,
  userName: string,
  phoneNumber: string,
  emailAddress: string,
  onUserNameChange: (input: string) => void
  onPhoneNumberChange: (input: string) => void
  onEmailAddressChange: (input: string) => void
  onScannerStart: () => void
  onScannerSuccess: (e : Event) => void
  onScannerCancel: () => void
}

interface State {
}

// TODO Magic numbers
const styles = StyleSheet.create({
  textInputField: {
    flex: 1,
    width: '80%'
  },
  block: {
    marginBottom: "15%"
  },
  icon: {
    margin: "20%"
  },
  iconContainer: {
    height: 55,
    width: 55,
    borderRadius: 35,
    backgroundColor: JolocomTheme.palette.primaryColor,
    alignItems: 'center',
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.8,
    shadowRadius: 2,
    elevation: 8,
  },
  actionButtonContainer: {
    position: 'absolute',
    right: '3%',
    bottom: '5%',
    alignItems: 'flex-end'
  },
  centerText: {
    flex: 1,
    fontSize: 18,
    padding: 32,
    color: '#777',
  },
  textBold: {
    fontWeight: '500',
    color: '#000',
  },
  buttonText: {
    fontSize: 21,
    color: 'rgb(0,122,255)',
  },
  buttonTouchable: {
    padding: 16,
  }
})

// TODO SFC?
export class IdentityComponent extends React.Component<Props, State> {
  render() {
    let renderContent
    if (this.props.scanning) {
      renderContent = (
        <QRcodeScanner
          onScannerSuccess={this.props.onScannerSuccess}
          onScannerCancel={this.props.onScannerCancel}
        />
      )
    } else {
      renderContent = (
        <Block>
          <Block style={ styles.block }>
            <TextInput
              style={ styles.textInputField }
              placeholder='Please enter your name'
              onChangeText={ this.props.onUserNameChange }
              value={ this.props.userName}
            />
            <TextInput
              style={ styles.textInputField }
              placeholder='Please enter your phone number'
              onChangeText={ this.props.onPhoneNumberChange}
              value={ this.props.phoneNumber}
            />
            <TextInput
              style={ styles.textInputField }
              placeholder='Please enter your email address'
              onChangeText={ this.props.onEmailAddressChange}
              value={ this.props.emailAddress}
            />
          </Block>
          <Block style={ styles.actionButtonContainer }>
            <TouchableOpacity
              style={ styles.iconContainer }
              onPress={ this.props.onScannerStart }>
              <Icon
                style={ styles.icon }
                size={ 30 }
                name="qrcode-scan"
                color="white"
              />
          </TouchableOpacity>
        </Block>
      </Block>
      )
    }

    return (
      <Container>
        {renderContent}
      </Container>
    )
  }
}
