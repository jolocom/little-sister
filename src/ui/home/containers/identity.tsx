import React from 'react'
import { IdentityComponent } from 'src/ui/home/components/identity'
import { connect } from 'react-redux'
import { RootState } from 'src/reducers/'

interface ConnectProps { }

interface Props extends ConnectProps {}

interface State {
  userName: string,
  phoneNumber: string,
  emailAddress: string,
  scanning: boolean
}

export class IdentityContainer extends React.Component<Props, State> {

  state = {
    userName: '',
    phoneNumber: '',
    emailAddress: '',
    scanning: false
  }

  componentDidMount() {
  }
  private onUserNameChange= (userName: string) : void => {
    this.setState({ userName })
  }

  private onPhoneNumberChange= (phoneNumber: string) : void => {
    this.setState({ phoneNumber })
  }

  private onEmailAddressChange= (emailAddress: string) : void => {
    this.setState({ emailAddress })
  }

  private onScannerStart = () : void => {
    this.setState({ scanning: true })
  }

  private onScannerCancel = () : void => {
    this.setState({ scanning: false })
  }

  // TODO Typings on E, event is not enough
  private onScannerSuccess = (e : Event) : void => {
    this.setState({ scanning: false })
  }

  render() {
    return (
    <IdentityComponent
      scanning={ this.state.scanning }
      userName= {this.state.userName}
      phoneNumber= { this.state.phoneNumber }
      emailAddress= { this.state.emailAddress }
      onUserNameChange= { this.onUserNameChange }
      onPhoneNumberChange= { this.onPhoneNumberChange }
      onEmailAddressChange= { this.onEmailAddressChange }
      onScannerStart={ this.onScannerStart }
      onScannerSuccess={ this.onScannerSuccess }
      onScannerCancel={ this.onScannerCancel }
     />
    )
  }
}

const mapStateToProps = (state: RootState) => {
  return {
  }
}

const mapDispatchToProps = (dispatch: (action: Function) => void) => {
  return {
  }
}

export const Identity = connect(mapStateToProps, mapDispatchToProps)(IdentityContainer)
