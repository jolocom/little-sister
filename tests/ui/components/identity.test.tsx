import React from 'react'
import { IdentityComponent } from 'src/ui/home/components/identity'
import { shallow } from 'enzyme'

describe('Identity component', ()=> {
  const COMMON_PROPS = {
    scanning: false,
    userName: '',
    phoneNumber: '',
    emailAddress: '',
    onUserNameChange: () => null,
    onPhoneNumberChange: () => null,
    onEmailAddressChange: () => null,
    onScannerStart: () => null,
    onScannerSuccess: () => null,
    onScannerCancel: () => null
  } 
  it('matches the snapshot on render', () => {

    const rendered = shallow(<IdentityComponent {...COMMON_PROPS}/>)
    expect(rendered).toMatchSnapshot()
  })
})
