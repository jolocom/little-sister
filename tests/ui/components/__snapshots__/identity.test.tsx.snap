// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Identity component matches the snapshot on render 1`] = `
<Container>
  <Block>
    <Block
      style={
        Object {
          "marginBottom": "15%",
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Please enter your name"
        style={
          Object {
            "flex": 1,
            "width": "80%",
          }
        }
        value=""
      />
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Please enter your phone number"
        style={
          Object {
            "flex": 1,
            "width": "80%",
          }
        }
        value=""
      />
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Please enter your email address"
        style={
          Object {
            "flex": 1,
            "width": "80%",
          }
        }
        value=""
      />
    </Block>
    <Block
      style={
        Object {
          "alignItems": "flex-end",
          "bottom": "5%",
          "position": "absolute",
          "right": "3%",
        }
      }
    >
      <TouchableOpacity
        activeOpacity={0.2}
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#942f51",
            "borderRadius": 35,
            "elevation": 8,
            "height": 55,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.8,
            "shadowRadius": 2,
            "width": 55,
          }
        }
      >
        <Icon
          allowFontScaling={false}
          color="white"
          name="qrcode-scan"
          size={30}
          style={
            Object {
              "margin": "20%",
            }
          }
        />
      </TouchableOpacity>
    </Block>
  </Block>
</Container>
`;
